### DISCUSSION TITLE
Why is compressionQuality parameter ignored in Img32.Fmt.PNG?

### DISCUSSION THREAD

**Discussion by idw-git** [AngusJohnson/Image32#114] [General]
Hi Angus, I recently installed your latest version of Image32 and am wondering why Img32.Fmt.PNG says on line 28 that "SaveToStream: the compressionQuality parameter is ignored here"? Does this have any specific reason or was it actually something postponed for later implementation? Of course, it's easy enough to apply compressionQuality in SaveToStream for the non-FPC section by myself, but I'm still wondering.

**Comment by AngusJohnson** [comment]
The generic (ancestor) `TImageFormat` class from which `TImageFormat_PNG` class derives is also the ancestor of `TImageFormat_JPG`. PNG compression is loseless, unlike JPG compression, so the amount that PNG images can be compressed is generally fixed *. 


I had a choice of **overloading** or **overriding** the `TImageFormat.SaveToStream` method,  and I chose the latter option.

*Edit: Nevertheless, I concede that PNG images are still compressed using ZLIB compression, though most usually using the default level (6 or 7 depending where you look 😉).

**Comment by idw-git** [comment]
Ok, so you're rather disscussing the naming of this parameter instead of an option for using it as a compressionLevel parameter for PNGs. But since it's just a plain integer type anyway it could easily be used as an open "Option" parameter for different overrides.

Here's what I did: First of all add `png.CompressionLevel := compressionQuality;` in Img32.Fmt.PNG's "SaveToStream" method, then create a derived "TImageFormat_PNGX" like so:

```TImageFormat_PNGX = class(TImageFormat_PNG)
  private
    class var FCompressionLevel: TCompressionLevel;
  public
    procedure SaveToStream(stream: TStream; img32: TImage32; CompressionLevel: Integer); override;
    class property CompressionLevel: TCompressionLevel read FCompressionLevel write FCompressionLevel;
  end;

implementation

procedure TImageFormat_PNGX.SaveToStream(stream: TStream; img32: TImage32; CompressionLevel: Integer);
begin
  inherited SaveToStream(stream, img32, FCompressionLevel);
end;

initialization
  TImage32.RegisterImageFormatClass('PNG', TImageFormat_PNGX, cpHigh);

end.```

**Comment by AngusJohnson** [comment]
I don't believe your code above would work because the `CompressionLevel` parameter is still ignored.
Nevertheless I've modified the code so the compression level is no longer ignored 😁.

**Reply by idw-git** [reply]
My code does work: It would not work if using `TImageFormat_PNGX.SaveToStream()` directly (according to my tests), but if using `TImage32.SaveToStream()` it does, provided the `CompressionLevel` property has been set using an initialization routine like this: `TImageFormat_PNGX(GetImageFormatClass(PNG)).CompressionLevel := PNGCompressionLevel` in a component derived from `TImage32`.

It is still required after your latest fix and I think I now understand why: While `TImage.SaveToFile()` _has_ a `quality` parameter `TImage.SaveToStream()` does _not_: So there's currenly no way to pass this parameter to the `TImageFormat.SaveToStream()` proc while you can certainly pass it to `TImageFormat.SaveToFile()`.

**Reply by AngusJohnson** [reply]
> My code does work

As far as I can see in your code above, the CompressionLevel **param** is ignored because you're using the CompressionLevel **property** inside your `SaveToStream` method.

> TImage.SaveToFile() has a quality parameter TImage.SaveToStream() does not

OK, I will investigate 😁.
